### ISSUE TITLE
Устранение избыточных нейронов

### ISSUE THREAD

**Issue by igsha** [helenross636/magic_season#9]
Делается после #7.

Входные и выходные веса похожих нейронов нужно сложить. После операции сложения нужно запустить сеть и посмотреть на точность сети.

**Comment by helenross636** [comment]
После внесений изменений в модель надо заново выполнять model.fit(), чтобы ее «запустить»? Или надо еще делать model.compile()?

**Comment by igsha** [comment]
Делать `fit` и `compile` не нужно. Нужно делать только `evaluate`, чтобы оценить тестовую точность.

**Comment by helenross636** [comment]
Еще вопрос: как именно "складывать" входные и выходные веса нейронов? понятно, что надо изменять входные веса скрытого слоя. Но у них же разный формат, например, (30, 784) и (30, 10).

**Comment by igsha** [comment]
Входные веса складываются нейрона с входными весами другого нейрона. То же самое проделать с выходными весами. В итоге 2 нейрона превратятся в один. Не нужно складывать вход с выходом.

**Comment by igsha** [comment]
Исходя из ковариационной матрицы (модуль, диагональ обнулена), построить маску выходных весов (следить, чтобы количество строк совпадало с количеством нейронов):
```python
w = model.layers[2].kernel
print(w.shape)
am = np.abs(np.corrcoef(w))
np.fill_diagonal(am, 0)
mask = am > 0.9
```
Теперь по этой маске нужно искать "похожие" нейроны:
```python
similar = {}
for idx, k in enumerate(mask):
    nk = k.nonzero()[0]
    if len(nk) != 0:
        similar[idx] = nk[np.argmax(am[idx, nk])]

print(len(similar))
```
Мы получим индексы похожих нейронов в терминах корелляции выходных весов. Необходимо их попечатать и попытаться понять, как из них можно создать новую сеть.

**Comment by helenross636** [comment]
Пришлось понизить порог до 0.5:
https://github.com/helenross636/magic_season/blob/main/%23Task7.ipynb

**Comment by igsha** [comment]
> Пришлось понизить порог до 0.5: https://github.com/helenross636/magic_season/blob/main/%23Task7.ipynb

 Это, наверное, не к этому заданию. В этом задании нужно удалять нейроны и модифицировать веса.

**Comment by helenross636** [comment]
Не правильно подписала. В самом конце, после надписи "#TASK 8". там по маске искались похожие нейроны. И вместо
`mask = am > 0.9`
взяла 
`mask = am > 0.5`

**Comment by igsha** [comment]
Нужно дообучить модели получше. Для больших сетей доведите до 96-98%

**Comment by helenross636** [comment]
Получилось поднять порог до 0.7:
https://github.com/helenross636/magic_season/blob/main/%23Task9.ipynb

**Comment by helenross636** [comment]
По поводу сложения: надо брать любые пары и их складывать? и для пар надо выбирать те нейроны, у которых веса  имеют значения ниже порога?

**Comment by igsha** [comment]
Те нейроные, которые есть в словаре `similar` - это и есть пары похожих.

**Comment by helenross636** [comment]
Попробовала преобразовать входные и выходные веса похожих нейронов и запустить снова сети. 
Только не знаю, надо ли было как-то удалять из каждой пары по нейрону. В итоге просто складывала пары, образуя одинаковые нейроны. 

Если судить по результатам, то только наибольшая модель хоть как-то приблизилась по точности к изначальной, хотя и она не достигла его. 
https://github.com/helenross636/magic_season/blob/main/%23Task9.ipynb

**Comment by igsha** [comment]
Да, нужно удалять старые нейроны. Получается каждая пара заменяется одним нейроном. Лучше создать сеть заново и новые веса установить. Исследуйте пока только самую большую сеть, остальные не нужно смотреть.

**Comment by helenross636** [comment]
Я действительно в последний раз забыла указать seed, когда на другом устройстве пришлось заново обучить модели, чтобы получить готовые. (но до этого всегда были одни и те же). Поэтому еще раз проверила на первоначальных моделях результат сложения нейронов и точность подучилось почти равная исходным моделям (другие модели также сохранены): 
![image](https://user-images.githubusercontent.com/47567451/172475024-38a82aa9-279c-4553-953b-2db805d80e4e.png)

Далее попробовала создать новую модель, без повторяющихся нейронов. Однако, при этом было обнаружено, что некоторые пары (их не очень много, но они есть) пересекаются по одному из нейронов (пример приведен в In [96]). В итоге было удалено количество нейронов равное половине от общего числа пар.
После создания, передачи весов и запуска новой (назовем ее "обрезанной" или "model_CUT") модели замечено, что точность сильно упала (если нигде нет ошибок).
![image](https://user-images.githubusercontent.com/47567451/172475799-5edb696b-3442-4e81-958d-b90cd44f6158.png)

https://github.com/helenross636/magic_season/blob/main/%23Task9.ipynb

**Comment by igsha** [comment]
Что-то не то. Сейчас задачу создам другую. Её нужно проделать до этой.
